src/allotropy/parsers/thermo_fisher_nanodrop_eight/nanodrop_eight_parser.py

```diff
@@ -0,0 +1,272 @@
+from typing import Optional, TypeVar
+import uuid
+
+import pandas as pd
+
+from allotropy.allotrope.models.shared.definitions.custom import (
+    TQuantityValueMicrogramPerMicroliter,
+    TQuantityValueMicrogramPerMilliliter,
+    TQuantityValueMilliAbsorbanceUnit,
+    TQuantityValueMilligramPerMilliliter,
+    TQuantityValueNanogramPerMicroliter,
+    TQuantityValueNanogramPerMilliliter,
+    TQuantityValueNanometer,
+    TQuantityValuePicogramPerMilliliter,
+)
+from allotropy.allotrope.models.shared.definitions.definitions import (
+    InvalidJsonFloat,
+    JsonFloat,
+    TQuantityValue,
+)
+from allotropy.allotrope.models.spectrophotometry_benchling_2023_12_spectrophotometry import (
+    CalculatedDataAggregateDocument,
+    CalculatedDataDocumentItem,
+    DataSourceAggregateDocument1,
+    DataSourceDocumentItem,
+    DataSystemDocument,
+    DeviceSystemDocument,
+    MeasurementAggregateDocument,
+    Model,
+    SampleDocument,
+    SpectrophotometryAggregateDocument,
+    SpectrophotometryDocumentItem,
+    UltravioletAbsorbancePointDetectionDeviceControlAggregateDocument,
+    UltravioletAbsorbancePointDetectionDeviceControlDocumentItem,
+    UltravioletAbsorbancePointDetectionMeasurementDocumentItems,
+)
+from allotropy.constants import ASM_CONVERTER_NAME, ASM_CONVERTER_VERSION
+from allotropy.named_file_contents import NamedFileContents
+from allotropy.parsers.thermo_fisher_nanodrop_eight.nanodrop_eight_reader import (
+    NanoDropEightReader,
+)
+from allotropy.parsers.vendor_parser import VendorParser
+
+T = TypeVar("T")
+
+CONCENTRATION_UNIT_TO_TQUANTITY = {
+    "ug/ul": TQuantityValueMicrogramPerMicroliter,
+    "ug/ml": TQuantityValueMicrogramPerMilliliter,
+    "mg/ml": TQuantityValueMilligramPerMilliliter,
+    "ng/ul": TQuantityValueNanogramPerMicroliter,
+    "ng/ml": TQuantityValueNanogramPerMilliliter,
+    "pg/ul": TQuantityValuePicogramPerMilliliter,
+}
+
+
+def _get_str(data_frame: pd.DataFrame, row: int, column: str) -> Optional[str]:
```

Comment:
I don't think we should allow this method to return `None`.

If `column` isn't found, we should throw.

If the value is na, we should return _some_ string. This will ensure that we have strings to operate on wherever this is called.

Reply:
Some fields will accept None as a value and in that case will not be represented in the generated ASM.  Returning None simplifies the logic in those cases.  Would it make sense to have a method like _get_str that returns a atr or throws an error and a separate method like _get_str_or_none that acts like the current _get_str?

Reply:
Yeah, renaming this to `_get_str_or_none` and having `_get_str` return a `str` or throw sounds good. The latter should be used in all the places where you were calling `str` on an `Optional[str]`.

Reply:
for _get_str on a field that has None, do you have recommendation on error type and message?  Should this be AllotropeConversionError or something more general like ValueError?  As a message I was thinking something like "Error in _get_str:  Expecting str value but found None."

Reply:
or maybe TypeError: _get_str does not support 'NoneType' values

Reply:
We're trying to return `AllotropeConversionError` for all exceptions thrown from allotropy. The most common pattern for null validation seems to be `assert_not_none`, so I'd just that.